### ISSUE TITLE
Use nix-tree-sitter for the parser

### ISSUE THREAD

**Issue by domenkozar** [haskell-nix/hnix#508]
Currently, parser performance is a big bottleneck, I suggest using https://github.com/cstrahan/tree-sitter-nix (look at [canonix](https://github.com/hercules-ci/canonix/) how to wire it up).

As @ryantrinkle wants to have GHCJS support I suggest tree-sitter is default parser, but megaparsec is used when GHCJSing.

**Comment by Ericson2314** [comment]
Interesting. I would love to teach `happy` those tricks so we can have our all-haskell GHCJS-supporting cake and savor it nice and incrementally too.

**Comment by domenkozar** [comment]
There is https://github.com/tree-sitter/tree-sitter/pull/321 but since whole editor space is moving in this direction, I'm sure it's also open for GHCJS at some point (not sure how well it integrated with wasm atm).

**Comment by vmchale** [comment]
I assume `happy`/`alex` would fix performance just fine? If you're using megaparsec to parse a programming language that would certainly explain why parsing is the bottleneck.

**Comment by Ericson2314** [comment]
Wish granted! @alanz once started on https://github.com/alanz/happy/tree/repetitive-rebased, demoed in https://github.com/alanz/incremental-play .

**Comment by Ericson2314** [comment]
I am now a happy maintainer :thinking: :thinking: :thinking:

**Comment by alanz** [comment]
@Ericson2314 what are you using for this? Anything we can retrofit to GHC in time?

**Comment by Ericson2314** [comment]
@alanz I am not currently using your branch for anything, but I sure would like to! My plan is to try to simplify the build system starting with https://github.com/simonmar/happy/issues/175, to "free up some complexity budget". Ultimately I want to see tree-sitter-like stuff with Reflex.

**Comment by Anton-Latukha** [comment]
Overall use of the 3rd party engine for parser seems good.

John wrote the project in recursive schemes. Which is a pretty unique property.
But it is also the central weakness - it raises the bar for contributors too high - so nobody can come in and if they come they can not understand the project and so - move a needle in the project.

I had several occasions when people said that HNix was too strange/heavy code to grasp and that they went away, others spend year or more practicing before they started to understand something.

One of the central points that I see - is that project needs to become more standard, to try to deliver more simply hard parts so would be approachable for people. I thought that it is possible refining the recursive schemes approaches with using libs for it where it applies, using tree-sitter seems as a much radical but probably very effective way, it would probably remove most of the recursive scheme's use from the project.

**Comment by Ericson2314** [comment]
Oh @Anton-Latukha please keep the recursion schemes I am a huge fan of them and think they are in fact the best part of hnix! By contrast, I viewed this as a good challenge for `happy` someday to better support recursion schemes with something like "grammar combinators".

**Comment by Anton-Latukha** [comment]
I also think that rec schemes are in the most proper place, it is a classic use for them. And also I already took privately some critique that today there are more modern approaches to apply rec schemes.

I am still a newb, I just try to move a needle in the projects. Learning rec schemes - can do, refactor them with libs - can do. Refactor such live project to the tree-sitter - no not this year of exp.

**Comment by Ericson2314** [comment]
Yeah I don't see this issue is really a "request" of hnix, but rather a challenge to other projects to get the situation good enough where hnix can easily have an incremental parser. Until then, it can just sit.
